hudi-client/hudi-client-common/src/test/java/org/apache/hudi/table/TestBaseTableServicePlanActionExecutor.java

```diff
@@ -0,0 +1,299 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class TestBaseTableServicePlanActionExecutor {
+
+  @TempDir
+  File tempFile;
+  private static final StorageConfiguration<?> CONF = getDefaultStorageConf();
+  private final HoodieEngineContext context = new HoodieLocalEngineContext(CONF);
+
+  /**
+   * Clustering commit : cl
+   * Commit : c
+   * |requestTime<---commit type--->completionTime|(written partitions in current commit)
+   * ---------------------------------------------------------------------------------------------------------------------------------------------> timeline
+   * |0<-c1->1|(part=0000)
+   *           |2<------cl2------>4|
+   *                        |3<-----------c3--------->6|(part=0003)
+   *                                         |5<-----------c4--------->7|(part=0005)
+   *                                                                                |8<-------------------c5----------------->10|(part=0008)
+   *                                                                                                             | start incr clustering at 9
+   */
+  @Test
+  public void testExecutorWithMultiWriter() throws Exception {
+    String tableName = "testTable";
+    String tablePath = tempFile.getAbsolutePath() + StoragePath.SEPARATOR + tableName;
+    HoodieTableMetaClient metaClient = HoodieTestUtils.init(
+        HoodieTestUtils.getDefaultStorageConf(), tablePath, HoodieTableType.COPY_ON_WRITE, tableName);
+    HoodieWriteConfig writeConfig = HoodieWriteConfig.newBuilder().withPath(tablePath)
+        .withIndexConfig(HoodieIndexConfig.newBuilder().withIndexType(HoodieIndex.IndexType.INMEMORY).build())
+        .withMarkersType("DIRECT")
+        .build();
+
+    // <requestTime, CompletionTime>
+    HashMap<String, String> instants = new HashMap<>();
+    instants.put("0000", "0001"); // commit 1
+    instants.put("0002", "0004"); // clustering 1
+    instants.put("0003", "0006"); // commit 2
+    instants.put("0005", "0007"); // commit 3
+    instants.put("0008", "0010"); // commit 4
+
+    prepareTimeline(metaClient, instants);
+
+    DummyTableServicePlanActionExecutor executor = new DummyTableServicePlanActionExecutor(context, writeConfig, getMockHoodieTable(metaClient), "0009");
+    Set<String> incrementalPartitions = (Set<String>)executor.getIncrementalPartitions(TableServiceType.CLUSTER).getRight();
+
+    assertEquals(incrementalPartitions.size(), 2);
+    assertTrue(incrementalPartitions.contains("0003"));
+    assertTrue(incrementalPartitions.contains("0005"));
+    assertFalse(incrementalPartitions.contains("0008"));
+  }
+
+  @Test
+  public void testGetPartitionsFallbackToFullScan() throws Exception {
```

Comment:
@danny0405 This UT is test fall back to full scan when table service instant is archived.